File: loa/ring/neopixel.py
"""Ring — hardware layer. WS2812B over SPI0, Raspberry Pi 5 safe.

Why SPI and not rpi_ws281x: the PyPI rpi_ws281x 5.0.0 wheel ships no compiled
binary for the RP1 chipset ("Hardware revision is not supported"), and the
Adafruit path needs Blinka. SPI is the Pi-5-safe, dependency-light route:
GPIO10 (MOSI) at 3.2MHz, four SPI bits per WS2812 bit.
"""
ONE  = bytes([0b1110])   # WS2812 "1"  ~937ns high, 1.25us total
ZERO = bytes([0b1000])   # WS2812 "0"  ~937ns low,  1.25us total


def _ws_byte(v: int) -> bytes:
    out = bytearray()
    for bit in (7, 6, 5, 4, 3, 2, 1, 0):
        out += ONE if (v >> bit) & 1 else ZERO
    return bytes(out)


def parse_color(value):
    """Normalize a color to an (r,g,b) tuple.

    Accepts:
      "#00FF00" / "#0F0"      hex string
      "0x00FF00"              hex string with prefix
      0x00FF00                int
      (0, 255, 0)             tuple or list of three ints

    Raises ValueError on anything else.
    """
    if isinstance(value, str):
        s = value.strip()
        if s.startswith("#"):
            s = s[1:]
        elif s.startswith("0x"):
            s = s[2:]
        if len(s) == 3:
            s = "".join(c * 2 for c in s)
        if len(s) != 6:
            raise ValueError(f"bad hex color: {value!r}")
        return (int(s[0:2], 16), int(s[2:4], 16), int(s[4:6], 16))
    if isinstance(value, (tuple, list)):
        if len(value) != 3:
            raise ValueError(f"color must be (r,g,b): {value!r}")
        return (int(value[0]), int(value[1]), int(value[2]))
    if isinstance(value, int):
        return ((value >> 16) & 0xFF, (value >> 8) & 0xFF, value & 0xFF)
    raise ValueError(f"unsupported color: {value!r}")


def grb(r: int, g: int, b: int) -> bytes:
    """Encode one pixel in WS2812 order (green, red, blue)."""
    return _ws_byte(g) + _ws_byte(r) + _ws_byte(b)


class Ring:
    """A WS2812B ring on SPI0.

    Args:
        num: number of LEDs (loa's ring is 24).
        bus, device: SPI bus/device (default 0,0 = /dev/spidev0.0).
        speed: SPI clock in Hz. 3.2MHz gives four SPI bits per WS2812 bit.
    """

    def __init__(self, num: int = 24, bus: int | None = None,
                 device: int | None = None, speed: int = 3_200_000):
        import spidev              # lazy: animations stay importable anywhere
        from .. import config
        cfg = config.load()        # loa.conf is the as-built truth
        self.num = num
        self.spi = spidev.SpiDev()
        self.spi.open(
            int(cfg.get("ring_bus", bus if bus is not None else 0)),
            int(cfg.get("ring_device", device if device is not None else 0)),
        )
        self.spi.max_speed_hz = int(cfg.get("ring_speed", speed))
        self.spi.mode = 0b00

    def show(self, frame) -> None:
        """Render one frame: an iterable of colors, len == self.num.

        Each color may be a tuple (r,g,b) or a hex string like "#00FF00".
        """
        buf = bytearray()
        for color in frame:
            r, g, b = parse_color(color)
            buf += grb(int(r), int(g), int(b))
        buf += b'\x00' * 24            # latch: 60us low
        self.spi.writebytes2(list(buf))
        self._publish(frame)

    #: Where the frames go. Set by the daemon (loa-ring) to put them on the
    #: topic; the driver itself stays ignorant of the wire, and a bench run with
    #: no daemon simply has no sink. The old path was a file in /dev/shm that a
    #: consumer read behind the publisher's back — a second data path wearing a
    #: mirror's clothes.
    frame_sink = None

    def _publish(self, frame) -> None:
        """Hand the exact display values (72B, 24 px RGB) to the sink."""
        sink = self.frame_sink
        if sink is None:
            return
        try:
            raw = bytearray()
            for color in frame:
                r, g, b = parse_color(color)
                raw += bytes((int(r), int(g), int(b)))
            sink(bytes(raw))
        except Exception:
            pass

    def fill(self, rgb) -> None:
        """Set every LED to one color — tuple, hex string, or int."""
        self.show([rgb] * self.num)

    def off(self) -> None:
        self.fill((0, 0, 0))

    def close(self) -> None:
        self.spi.close()

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()

File: rust/loa-ring/src/neopixel.rs
//! The WS2812B, over SPI, on the Pi 5.
//!
//! WS2812 timing is what this file is about. There is no "set brightness"
//! register: each LED bit is a pulse whose WIDTH is the data, and the strip
//! latches 60us of low at the end. rpi_ws281x cannot drive this chipset (its
//! wheel has no RP1 binary) and the Adafruit path needs Blinka, so the pulses
//! are made by SPI: four SPI bits per WS2812 bit at 3.2MHz.
//!
//! ```text
//!     1 -> 0b1110   ~937ns high, 1.25us total
//!     0 -> 0b1000   ~937ns low,  1.25us total
//! ```
//!
//! The clock is the contract. At 3.2MHz eight SPI bits are 2.5us, so four of
//! them are 1.25us — exactly one WS2812 bit period. Change the clock and the
//! colours change with it, which is why the speed is a setting and not a
//! constant in the loop.

use std::io::Write;

pub const LEDS: usize = 24;
pub const BYTES: usize = LEDS * 3;             // 72: 24 px RGB
const SPI_BYTE_PER_WS_BIT: usize = 1;
const LATCH: usize = 24;                       // 60us low at 3.2MHz

/// 3.2MHz: four SPI bits per WS2812 bit, which is where the numbers above
/// come from. Measured as the as-built truth in loa.conf on the body.
pub const DEFAULT_HZ: u32 = 3_200_000;

const ONE: u8 = 0b1110;
const ZERO: u8 = 0b1000;

/// One channel value -> eight SPI bytes. MSB first: the strip takes the high
/// bit of the colour first, so the order is part of the protocol, not a
/// preference.
///
/// Public because the wire format is the part of this port that fails
/// SILENTLY — bad timing raises nothing, it shows the wrong colours — so
/// `tests/wire_format.rs` pins it against the Python encoder's own bytes.
pub fn ws_byte(v: u8, out: &mut Vec<u8>) {
    for bit in (0..8).rev() {
        out.push(if (v >> bit) & 1 == 1 { ONE } else { ZERO });
    }
}

pub struct Ring {
    spi: spidev::Spidev,
}

impl Ring {
    pub fn open(path: &str, speed_hz: u32) -> Result<Self, String> {
        let mut spi = spidev::Spidev::open(path).map_err(|e| format!("{path}: {e}"))?;
        let opts = spidev::SpidevOptions::new()
            .bits_per_word(8)
            .max_speed_hz(speed_hz)
            .mode(spidev::SpiModeFlags::SPI_MODE_0)
            .build();
        spi.configure(&opts).map_err(|e| e.to_string())?;
        Ok(Ring { spi })
    }

    /// 72 bytes of RGB (r,g,b per LED) -> the strip. GRB on the wire: the
    /// WS2812B takes green first, which is a property of the part and not
    /// something the renderer has to know about.
    pub fn show(&mut self, rgb: &[u8]) -> Result<(), String> {
        if rgb.len() != BYTES {
            return Err(format!("frame is {} bytes, expected {BYTES}", rgb.len()));
        }
        let mut buf = Vec::with_capacity(BYTES * 8 * SPI_BYTE_PER_WS_BIT + LATCH);
        for px in rgb.chunks_exact(3) {
            ws_byte(px[1], &mut buf);          // G
            ws_byte(px[0], &mut buf);          // R
            ws_byte(px[2], &mut buf);          // B
        }
        buf.extend(std::iter::repeat_n(0u8, LATCH));   // latch: 60us low
        self.spi.write_all(&buf).map_err(|e| e.to_string())
    }

    pub fn off(&mut self) -> Result<(), String> {
        self.show(&[0u8; BYTES])
    }
}
